test/malli/distributive_test.cljc

```diff
@@ -0,0 +1,79 @@
+(ns malli.distributive-test
+  (:require [clojure.test :refer [are deftest is testing]]
+            [malli.core :as m]
+            [malli.impl.util :as miu]
+            [malli.registry :as mr]
+            [malli.transform :as mt]
+            [malli.util :as mu]))
+
+(def options {:registry (merge (mu/schemas) (m/default-schemas))})
+
+(deftest distributive-test
+  (is (= (m/form
+           (m/deref
+             [:merge
+              [:map [:x :int]]
+              [:multi {:dispatch :y}
+               [1 [:map [:y [:= 1]]]]
+               [2 [:map [:y [:= 2]]]]]]
+             options))
+         [:multi {:dispatch :y}
+          [1 [:map [:x :int] [:y [:= 1]]]]
+          [2 [:map [:x :int] [:y [:= 2]]]]]))
+  (is (= (m/form
+           (m/deref
+             [:merge
+              [:multi {:dispatch :y}
+               [1 [:map [:y [:= 1]]]]
+               [2 [:map [:y [:= 2]]]]]
+              [:map [:x :int]]]
+             options))
+         [:multi {:dispatch :y}
+          [1 [:map [:y [:= 1]] [:x :int]]]
+          [2 [:map [:y [:= 2]] [:x :int]]]]))
+
+  (is (= (m/form
+           (m/deref
+             [:merge
+              [:map [:x :int]]
+              [:orn
+               [1 [:map [:y [:= 1]]]]
+               [2 [:map [:y [:= 2]]]]]]
+             options))
+         [:orn
+          [1 [:map [:x :int] [:y [:= 1]]]]
+          [2 [:map [:x :int] [:y [:= 2]]]]]))
+  (is (= (m/form
+           (m/deref
+             [:merge
+              [:orn
+               [1 [:map [:y [:= 1]]]]
+               [2 [:map [:y [:= 2]]]]]
+              [:map [:x :int]]]
+             options))
+         [:orn
+          [1 [:map [:y [:= 1]] [:x :int]]]
+          [2 [:map [:y [:= 2]] [:x :int]]]]))
+
+  (is (= (m/form
+           (m/deref
+             [:merge
+              [:map [:x :int]]
+              [:or
+               [:map [:y [:= 1]]]
+               [:map [:y [:= 2]]]]]
```

Comment:
what happens if there is non-map here? gets replaced? I would not expect it to happen.

Reply:
maybe this should work only with `:multi` 🤔

Reply:
> I really like the approach here.

Seems powerful! I'm still grokking the implications and applicability myself...see below.

This is something Typed Clojure does to convert
```
(t/Merge '{:a t/Int} (t/U '{:b t/Int} '{:c t/Int})
```
to
```
(t/U '{:a t/Int :b t/Int} '{:a t/Int :c t/Int})
```
but such transformations are not so obviously applicable to Malli because of `m/parse` and error messages relying on named paths.

> What is the reasoning for this working with :or and :orn too?

I think all 3 are kinds of disjunctions so it seemed natural to at least try it. I also tried `:and` which might also be useful in some (non-map) contexts.

`:or` + `:map` is not performant but I was also curious if there are other schemas analogous to `:merge` for which `[<COMBINE> X [:or Y Z]]` => `[:or [<COMBINE> X Y] [<COMBINE> X Z]]` makes sense.

> what happens if there is non-map here? gets replaced? I would not expect it to happen.

`:merge` essentially wraps the distributed schema, so it inherits the semantics of merge in that case.

`:merge` does some weird things if you give it non-maps.

```clojure
(m/deref
  [:merge
   int?
   [:map [:x :int]]]
  options)
;=> [:map [:x :int]]

(m/deref
  [:merge
   [:map [:x :int]]
   int?]
  options)
;=> int?
```

Perhaps it should error in these cases? FWIW Typed Clojure implements `t/Merge` via a protocol [`AssocableType`](https://github.com/typedclojure/typedclojure/blob/a27632474c33a74aa26f6e4a23f136829857a2dc/typed/clj.checker/src/typed/clj/checker/assoc_utils.clj#L31) which knows how to `assoc` a particular entry onto a map. Merging then happens in an analogous way to `clojure.core/merge`, but on the type level (pour entries from right-to-left one at a time).

My first impression of `mu/merge` is that its implementation is too concrete and could be extensible.

Anyway, because of `:merge`'s behavior the distributive rule drops the schema on the left if only one of them are maps:

```clojure
(m/deref
  [:merge
   [:or
    int?
    boolean?]
   [:map [:x :int]]]
  options)
=> 
[:or [:map [:x :int]] [:map [:x :int]]]

(m/deref
  [:merge
   [:map [:x :int]]
   [:or
    int?
    boolean?]]
  options)
=>
[:or int? boolean?]
```

The idea behind this PR is to implement these distributive rules as abstractly as possible to open the door to other non-map applications, i.e.:

```clojure
[<*> X [<+> Y Z]]
=>
[<+> [<*> X Y] [<*> X Z]]

[<*> [<+> Y Z] X]
=>
[<+> [<*> Y X] [<*> Z X]]
```

Going with that theme, here `:merge` is a kind of `<*>` schema that doesn't quite fit in the abstraction because of its own inconsistent semantics around non-maps.

> maybe this should work only with :multi 🤔

For maps it seems like the most obvious winning combination. Still looking for other applications, but the way it's implemented we can always add others later.

We could also support wrapper schemas like `:schema` such as

```clojure
[:merge
 X
 [:schema Y Z]]
```

using the same approach as `m/-function-schema?`.

Reply:
> It would be a breaking change to start failing with. non-entry schemas.

Agreed. A non-starter.

> I was following the [meta-merge](https://github.com/weavejester/meta-merge) functionality here:

It seems like it's all working as intended then. Applying the distributive rule would only be surprising if you expected `:merge` to have different semantics in the first place.